### PR TITLE
Elasticsearch/users

### PR DESCRIPTION
**Pull request by Ayoush**
**Why:-** 
1) User Search and Widget Search is required to search for users and widget across the platform.
2) Roles are there to distinguished various roles at all level in the organisation which have different access.
3) User needs to be separated out based on whether they are invited or not.

**Associations:-**
1) Users will have the role_id which will help distinguish between various roles.

**Tickets:-**
1) S5U1T4 (4pt) | BE: API to list all members with pagination.
2) S5U5T1 (2pt) | BE: API to modify the role of the user.
3) S5U1T5 (4pt) | BE: Search API for all users by name.
4) S5U1T3 (2pt) | BE: Keep a field in user to track which user invited him/her.
5) S5U1T2 (2pt) | BE: Modify user table to accommodate roles.
6) S5U1T1 (2pt) | BE: Create seed and create roles table to handle roles.

**Changes:-**
1) Added role seed and modified user seed to accommodate role_id.
2) Modified User to accommodate Role and is_invited field.
3) User and Widget Search API is added which will search directly from elastic-search.
4) Added test-case

### CODE REVIEW

File: apps/acqdat_api/lib/acqdat_api_web/router.ex [hunk 1]
```diff
@@ -41,6 +41,10 @@ defmodule AcqdatApiWeb.Router do
     resources "/widget", Widgets.WidgetController,
       only: [:create, :update, :delete, :index, :show]
 
+    get "/search_widgets", Widgets.WidgetController, :search_widget
+    get "/search_users", Widgets.UserController, :search_users
+    get "/index_users", Widgets.UserController, :index_users
```
Comment (chandradot99) [inline apps/acqdat_api/lib/acqdat_api_web/router.ex:46]: @Ayoush why is this listing of users and searching is inside the widgets module. Are these users related to the widgets ??

Comment (Ayoush) [reply apps/acqdat_api/lib/acqdat_api_web/router.ex:46]: yes

File: apps/acqdat_api/lib/acqdat_api_web/router.ex [hunk 2]
```diff
@@ -28,6 +28,16 @@ defmodule AcqdatApiWeb.Router do
     post "/validate-token", AuthController, :validate_token
     post "/sign-out", AuthController, :sign_out
 
+    resources "/org", OrganisationController, only: [:show] do
+      get "/users/search", UserController, :search_users
+
+      resources "/users", UserController, only: [:show, :update, :index] do
```
[the remark this answers is not in the document]
Comment (Ayoush) [reply apps/acqdat_api/lib/acqdat_api_web/router.ex:34]: is there another "/orgs/:org_id" and why you have  "org_id" defined explicitly if you just have used "/org" scope it will do the same.
What do you say ?

File: apps/acqdat_api/test/acqdat_api_web/controllers/user_controller_test.exs [hunk 3]
```diff
@@ -7,43 +7,88 @@ defmodule AcqdatApiWeb.UserControllerTest do
...
-    setup do
-      org = insert(:organisation)
-      [org: org]
-    end
-
-    test "fails if authorization header not found", context do
-      %{org: org, conn: conn} = context
+    test "fails if authorization header not found", %{conn: conn} do
       bad_access_token = "avcbd123489u"
+      org = insert(:organisation)
```
Comment (arjun289) [inline apps/acqdat_api/test/acqdat_api_web/controllers/user_controller_test.exs:12]: Please move org creation also inside setup no need to insert it every time.

File: apps/acqdat_api/test/acqdat_api_web/controllers/user_widget_controller_test.exs [hunk 4]
```diff
@@ -7,22 +7,16 @@ defmodule AcqdatApiWeb.UserWidgetControllerTest do
...
-      org = insert(:organisation)
-      user = insert(:user)
-      [org: org, user: user]
-    end
-
-    test "user widget create", context do
-      %{org: org, user: user, conn: conn} = context
+    test "user widget create", %{conn: conn, user: user} do
       widget = insert(:widget)
+      org = insert(:organisation)
```
Comment (arjun289) [inline apps/acqdat_api/test/acqdat_api_web/controllers/user_widget_controller_test.exs:12]: moe org creation to setup

File: apps/acqdat_api/test/acqdat_api_web/controllers/user_widget_controller_test.exs [hunk 5]
```diff
@@ -7,22 +7,16 @@ defmodule AcqdatApiWeb.UserWidgetControllerTest do
...
 
       params = %{
         user_id: user.id,
         widget_id: widget.id
       }
 
-      conn = post(conn, Routes.user_widgets_path(conn, :create, org.id, params), %{})
+      conn = post(conn, Routes.user_path(conn, :create, org.id, params), %{})
```
Comment (arjun289) [inline apps/acqdat_api/test/acqdat_api_web/controllers/user_widget_controller_test.exs:19]: This path does not look good to me, it appears by looking it's for user creation. Although, it is for user widget association.

File: apps/acqdat_core/lib/acqdat_core/model/user.ex [hunk 6]
```diff
@@ -51,4 +51,15 @@ defmodule AcqdatCore.Model.User do
...
+
+  @doc """
+  Update a User.
+
+  Expects `user` and update parameters as the arguments
+  """
+  def update(%User{} = user, params) do
+    changeset = User.update_changeset(user, params)
+    {:ok, user} = Repo.update(changeset)
+    {:ok, user |> Repo.preload(:role)}
```
Comment (arjun289) [inline apps/acqdat_core/lib/acqdat_core/model/user.ex:63]: This doesn't take care of the `error` case.

File: apps/acqdat_api/test/acqdat_api_web/controllers/user_setting_controller_test.exs [hunk 7]
```diff
@@ -22,15 +22,23 @@ defmodule AcqdatApiWeb.UserSettingControllerTest do
         |> put_req_header("authorization", "Bearer #{bad_access_token}")
 
       data = %{}
-      conn = post(conn, Routes.user_settings_path(conn, :create, org.id, 1), data)
+
+      conn =
+        post(conn, Routes.organisation_user_user_setting_path(conn, :create, org.id, 1), data)
```
[the remark this answers is not in the document]
Comment (Ayoush) [reply apps/acqdat_api/test/acqdat_api_web/controllers/user_setting_controller_test.exs:27]: Done

File: apps/acqdat_api/test/acqdat_api_web/controllers/user_setting_controller_test.exs [hunk 8]
```diff
@@ -22,15 +22,23 @@ defmodule AcqdatApiWeb.UserSettingControllerTest do
...
     end
 
     test "fails if required params are missing", context do
       %{org: org, user: user, conn: conn} = context
 
-      conn = post(conn, Routes.user_settings_path(conn, :create, org.id, user.id), %{})
+      conn =
+        post(
+          conn,
+          Routes.organisation_user_user_setting_path(conn, :create, org.id, user.id),
```
[the remark this answers is not in the document]
Comment (Ayoush) [reply apps/acqdat_api/test/acqdat_api_web/controllers/user_setting_controller_test.exs:39]: Done

File: apps/acqdat_api/test/acqdat_api_web/controllers/user_setting_controller_test.exs [hunk 9]
```diff
@@ -53,7 +61,12 @@ defmodule AcqdatApiWeb.UserSettingControllerTest do
         data_settings: user_setting.data_settings
       }
 
-      conn = post(conn, Routes.user_settings_path(conn, :create, org.id, user.id), data)
+      conn =
+        post(
+          conn,
+          Routes.organisation_user_user_setting_path(conn, :create, org.id, user.id),
```
[the remark this answers is not in the document]
Comment (Ayoush) [reply apps/acqdat_api/test/acqdat_api_web/controllers/user_setting_controller_test.exs:67]: Done

File: apps/acqdat_api/test/acqdat_api_web/controllers/user_setting_controller_test.exs [hunk 10]
```diff
@@ -78,19 +91,27 @@ defmodule AcqdatApiWeb.UserSettingControllerTest do
         |> put_req_header("authorization", "Bearer #{bad_access_token}")
 
       data = %{}
-      conn = post(conn, Routes.user_settings_path(conn, :create, org.id, 1), data)
+
+      conn =
+        post(conn, Routes.organisation_user_user_setting_path(conn, :create, org.id, 1), data)
```
[the remark this answers is not in the document]
Comment (Ayoush) [reply apps/acqdat_api/test/acqdat_api_web/controllers/user_setting_controller_test.exs:96]: Done

File: apps/acqdat_api/lib/acqdat_api/elasticsearch.ex [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+  end
+
+  def user_indexing(page) do
+    page_size = String.to_integer(page)
+    {:ok, _return_code, hits} = get("/users/_search", size: page_size)
+    {:ok, hits.hits}
+  end
+
+  def search_widget(_, _, _retry = 0), do: :ignore
+  def search_user(_, _, _retry = 0), do: :ignore
```
[the remark this answers is not in the document]
Comment (Ayoush) [reply apps/acqdat_api/lib/acqdat_api/elasticsearch.ex:109]: Done

File: apps/acqdat_api/lib/acqdat_api/elasticsearch.ex [hunk 12]
```diff
@@ -0,0 +1,110 @@
...
+        doc: [
+          label: params.label,
+          uuid: params.uuid,
+          properties: params.properties,
+          category: params.category
+        ]
+      )
+    end
+
+    GenRetry.retry(update_function, retries: 3, delay: 10_000)
```
[the remark this answers is not in the document]
Comment (Ayoush) [reply apps/acqdat_api/lib/acqdat_api/elasticsearch.ex:31]: Done

File: apps/acqdat_api/lib/acqdat_api_web/controllers/role_controller.ex [hunk 13]
```diff
@@ -0,0 +1,24 @@
...
+  import AcqdatApiWeb.Validators.Role
+  alias AcqdatCore.Model.Role, as: RoleModel
+
+  def index(conn, params) do
+    changeset = verify_index_params(params)
+
+    case conn.status do
+      nil ->
+        {:extract, {:ok, data}} = {:extract, extract_changeset_data(changeset)}
+        {:list, role} = {:list, RoleModel.get_all(data)}
```
[the remark this answers is not in the document]
Comment (Ayoush) [reply apps/acqdat_api/lib/acqdat_api_web/controllers/role_controller.ex:13]: Done

File: apps/acqdat_api/test/acqdat_api_web/controllers/user_widget_controller_test.exs (hunk 5 above)
Comment (Ayoush) [reply apps/acqdat_api/test/acqdat_api_web/controllers/user_widget_controller_test.exs:19]: Done

File: apps/acqdat_api/test/acqdat_api_web/controllers/user_widget_controller_test.exs (hunk 4 above)
Comment (Ayoush) [reply apps/acqdat_api/test/acqdat_api_web/controllers/user_widget_controller_test.exs:12]: Done

Comment (Ayoush) [reply apps/acqdat_api/test/acqdat_api_web/controllers/user_widget_controller_test.exs:12]: Done

File: apps/acqdat_api/test/acqdat_api_web/controllers/user_controller_test.exs (hunk 3 above)
Comment (Ayoush) [reply apps/acqdat_api/test/acqdat_api_web/controllers/user_controller_test.exs:12]: Done

File: apps/acqdat_api/lib/acqdat_api_web/views/user_view.ex [hunk 14]
```diff
@@ -1,14 +1,29 @@
...
+      role_id: user_details.role_id,
+      role:
+        render_one(Map.from_struct(Repo.get(Role, user_details.role_id)), RoleView, "role.json"),
+      user_setting: render_one(user_details.user_setting, UserView, "user_setting.json"),
+      org:
+        render_one(
+          Map.from_struct(Repo.get(Organisation, user_details.org_id)),
+          OrganisationView,
+          "org.json"
+        )
```
[the remark this answers is not in the document]
Comment (Ayoush) [reply apps/acqdat_api/lib/acqdat_api_web/views/user_view.ex:26]: done

File: apps/acqdat_api/lib/acqdat_api_web/views/user_view.ex [hunk 15]
```diff
@@ -19,4 +34,34 @@ defmodule AcqdatApiWeb.UserView do
...
+      email: hits.email,
+      org_id: hits.org_id,
+      role_id: hits.role_id,
+      role: render_one(Map.from_struct(Repo.get(Role, hits.role_id)), RoleView, "role.json"),
+      org:
+        render_one(
+          Map.from_struct(Repo.get(Organisation, hits.org_id)),
+          OrganisationView,
+          "org.json"
+        )
```
[the remark this answers is not in the document]
Comment (Ayoush) [reply apps/acqdat_api/lib/acqdat_api_web/views/user_view.ex:64]: done

File: apps/acqdat_core/lib/acqdat_core/model/user.ex (hunk 6 above)
Comment (Ayoush) [reply apps/acqdat_core/lib/acqdat_core/model/user.ex:63]: Done

File: apps/acqdat_api/lib/acqdat_api_web/controllers/user_controller.ex [hunk 16]
```diff
@@ -22,4 +28,71 @@ defmodule AcqdatApiWeb.UserController do
...
+      nil ->
+        case UserModel.update(conn.assigns.user, params) do
+          {:ok, user} ->
+            ElasticSearch.update_users("users", user)
+
+            conn
+            |> put_status(200)
+            |> render("user_details.json", %{user_details: user})
+
+          {:error, digital_twin} ->
```
[the remark this answers is not in the document]
Comment (Ayoush) [reply apps/acqdat_api/lib/acqdat_api_web/controllers/user_controller.ex:60]: Done

File: apps/acqdat_api/lib/acqdat_api_web/controllers/widgets/widget_controller.ex [hunk 17]
```diff
@@ -123,12 +128,18 @@ defmodule AcqdatApiWeb.Widgets.WidgetController do
     end
   end
 
+  def search_widget(conn, %{"label" => label}) do
+    with {:ok, hits} <- ElasticSearch.search_widget("widgets", label, 3),
+         do: conn |> put_status(200) |> render("hits.json", %{hits: hits})
```
[the remark this answers is not in the document]
Comment (Ayoush) [reply apps/acqdat_api/lib/acqdat_api_web/controllers/widgets/widget_controller.ex:133]: you already did.

File: apps/acqdat_api/lib/acqdat_api_web/controllers/user_controller.ex [hunk 18]
```diff
@@ -1,8 +1,14 @@
 defmodule AcqdatApiWeb.UserController do
   use AcqdatApiWeb, :controller
   alias AcqdatApi.User
+  alias AcqdatCore.Model.Organisation, as: OrgModel
+  alias AcqdatCore.Model.User, as: UserModel
+  alias AcqdatApi.ElasticSearch
   import AcqdatApiWeb.Helpers
 
+  plug :load_org when action in [:search_users, :index]
+  plug :load_user when action in [:show, :update]
```
[the remark this answers is not in the document]
Comment (Ayoush) [reply apps/acqdat_api/lib/acqdat_api_web/controllers/user_controller.ex:10]: Done

File: apps/acqdat_api/lib/acqdat_api_web/controllers/user_controller.ex (hunk 16 above)
[the remark this answers is not in the document]
Comment (Ayoush) [reply apps/acqdat_api/lib/acqdat_api_web/controllers/user_controller.ex:60]: Done
